### PR TITLE
Populate and expose object for volcana magmaman spawning logic's data.

### PR DESCRIPTION
**Pull request by jaythebusinessgoose**

### CODE REVIEW

**Comment by Dregu** [comment]
Hmm if I add some lava in dwelling and then call spawn_effects from volcana to start the magmaman logic, it works great and the current lava tiles are added to magmaman_positions automatically, but if I start the logic first, then spawn some lava and add the MagmamanSpawnPositions manually, the game crashes when I unload the level.

File: src/game_api/state_structs.hpp [hunk 1]
```diff
@@ -522,6 +522,22 @@ class LogicSunChallenge : public Logic
...
+  public:
+    uint32_t x;
+    uint32_t y;
+    uint32_t timer;
+};
+
+class LogicMagmamanSpawn : public Logic
+{
+  public:
+    std::vector<MagmamanSpawnPosition> magmaman_positions;
```
[the remark this answers is not in the document]
Comment (Dregu) [reply src/game_api/state_structs.hpp:536]: Yeah that works.

**Comment by Dregu** [comment]
So this would also benefit from a `MagmamanSpawnPosition(x, y)` constructor that sets the timer to a random number using whatever limits and prng the game uses, or maybe even a function to create and add a spawn position to the vector and start the logic automatically in different themes.

File: src/game_api/state_structs.hpp [hunk 2]
```diff
@@ -522,6 +522,22 @@ class LogicSunChallenge : public Logic
...
+
+    virtual ~LogicMagmamanSpawn() = 0;
```
Comment (Mr-Auto) [inline src/game_api/state_structs.hpp:538]: can you remove this virtual, it's not correct, not sure why all the logic have a separate destructor written, it's not there in the game, will change the others in my new PR

**Review by Mr-Auto — CHANGES_REQUESTED** [review]

**Comment by Dregu** [comment]
We don't currently have another way to start this logic anywhere, but this works

```lua
if not state.logic.magmaman_spawn then
  state.level_gen.themes[THEME.VOLCANA]:spawn_effects()
end
local x, y = math.floor(state.level_gen.spawn_x), math.floor(state.level_gen.spawn_y)
spawn_liquid(ENT_TYPE.LIQUID_LAVA, x, y)
table.insert(state.logic.magmaman_spawn.magmaman_positions, MagmamanSpawnPosition:new(x, y))
```

**Comment by Mr-Auto** [comment]
Does `table.remove` also work?

**Comment by Dregu** [comment]
Oh yeah or clearing the whole table to stop the spawns.

**Pull request merged**